app/Libraries/Beatmapset/BeatmapsetMainRuleset.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+declare(strict_types=1);
+
+namespace App\Libraries\Beatmapset;
+
+use App\Models\Beatmap;
+use App\Models\Beatmapset;
+use Ds\Set;
+
+class BeatmapsetMainRuleset
+{
+    /** @var ?Set<string> $eligibleRulesets */
+    private ?Set $eligibleRulesets = null;
+
+    public function __construct(private Beatmapset $beatmapset)
+    {
+        $values = $beatmapset->eligible_main_rulesets;
+
+        if ($values !== null) {
+            $this->eligibleRulesets = new Set($values);
+        } else {
+            $this->populateEligibleRulesets();
+        }
+    }
+
+    /**
+     * Gets all the Rulesets that are eligible to be the main ruleset.
+     * This will additionally query the current beatmapset nominations if necessary.
+     *
+     * @return Set<string>
+     */
+    public function currentEligible(): Set
+    {
+        $mainRuleset = $this->mainRuleset();
+
+        return $mainRuleset === null ? $this->eligibleRulesets : new Set([$mainRuleset]);
+    }
+
+    public function currentEligibleSorted(): array
+    {
+        $array = $this->currentEligible()->toArray();
+        sort($array);
+
+        return $array;
+    }
+
+    private function baseQuery()
+    {
+        return $this->beatmapset->beatmaps()
+            ->groupBy('playmode')
+            ->select('playmode', \DB::raw('count(*) as total'))
+            ->orderBy('total', 'desc')
+            ->orderBy('playmode', 'asc');
+    }
+
+    private function mainRuleset(): ?string
+    {
+        if ($this->eligibleRulesets->count() === 1) {
+            return $this->eligibleRulesets->first();
+        }
+
+        // First mode with more nominations that others becomes main mode.
+        // Implicity implies that limited BN nominations becomes main mode.
+        $nominations = $this->beatmapset->beatmapsetNominations()->current()->get();
```

Comment:
looks like this needs an explicit order. In my dev db the new row somehow is returned first and messed things up

Reply:
either that or the logic needs some rework to not immediately finish on the first ruleset with single nomination...